lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,297 @@
+        Distance.cosine(t1, t2, axes: [2])
+
+      :chebyshev ->
+        Distance.chebyshev(t1, t2, axes: [2])
+    end
+  end
+
+  defnp p_conditional(distances, sigmas) do
+    p = Nx.exp(-distances / (2 * Nx.reshape(sigmas, {:auto, 1})) ** 2)
+    {n, _} = Nx.shape(p)
+    p = Nx.put_diagonal(p, Nx.broadcast(0, {n}))
+    p = p + Nx.Constants.epsilon(:f32)
+    p / Nx.sum(p, axes: [1], keep_axes: true)
+  end
+
+  defnp perplexity(condition_matrix) do
+    exponent = -Nx.sum(condition_matrix * Nx.log2(condition_matrix), axes: [1])
+    2 ** exponent
+  end
+
+  defnp find_sigmas(distances, target_perplexity) do
+    {n, _} = Nx.shape(distances)
+    sigmas = Nx.broadcast(Nx.tensor(0, type: Nx.type(distances)), {n})
+
+    {sigmas, _, _, _} =
+      while {sigmas = sigmas, distances = distances, target_perplexity, i = 0}, i < n do
+        distances_i = Nx.take(distances, i)
+
+        sigmas =
+          Nx.indexed_put(
+            sigmas,
+            Nx.reshape(i, {1, 1}),
+            Nx.new_axis(binary_search(distances_i, target_perplexity), -1)
+          )
+
+        {sigmas, distances, target_perplexity, i + 1}
+      end
+
+    sigmas
+  end
+
+  defnp binary_search(distances, target_perplexity, opts \\ []) do
+    opts = keyword!(opts, tol: 1.0e-5, max_iters: 100, low: 1.0e-20, high: 1.0e4)
+    {low, high, max_iters, tol} = {opts[:low], opts[:high], opts[:max_iters], opts[:tol]}
+
+    {low, high, _, _, _, _, _, _} =
+      while {
+              low = low,
+              high = high,
+              max_iters,
+              tol,
+              perplexity_val = Nx.Constants.infinity(:f32),
+              distances,
+              target_perplexity,
+              i = 0
+            },
+            i < max_iters and Nx.abs(perplexity_val - target_perplexity) > tol do
+        mid = (low + high) / 2
+
+        condition_matrix = p_conditional(distances, Nx.new_axis(mid, 0))
+        perplexity_val = perplexity(condition_matrix) |> Nx.reshape({})
+
+        high_to_mid? =
+          cond do
+            perplexity_val > target_perplexity ->
+              1
+
+            true ->
+              0
+          end
+
+        low = if high_to_mid?, do: low, else: mid
+        high = if high_to_mid?, do: mid, else: high
+
+        {low, high, max_iters, tol, perplexity_val, distances, target_perplexity, i + 1}
+      end
+
+    (high + low) / 2
+  end
+
+  defnp q_joint(y, metric) do
+    distances = pairwise_dist(y, metric)
+    n = Nx.axis_size(distances, 0)
+    inv_distances = 1 / (1 + distances)
+    inv_distances = Nx.put_diagonal(inv_distances, Nx.broadcast(0, {n}))
+    inv_distances / Nx.sum(inv_distances)
+  end
+
+  defnp gradient(p, q, y, metric) do
+    pq_diff = p - q
+    y_diff = Nx.new_axis(y, 1) - Nx.new_axis(y, 0)
+    distances = pairwise_dist(y, metric)
+
+    inv_distances = 1 / (1 + distances)
+
+    ((4 * (Nx.new_axis(pq_diff, 2) * y_diff * Nx.new_axis(inv_distances, 2)))
+     |> Nx.sum(axes: [1])) * 2
```

Comment:
```suggestion
    pq_diff = Nx.new_axis(p - q, 2)
    y_diff = Nx.new_axis(y, 1) - Nx.new_axis(y, 0)
    distances = pairwise_dist(y, metric)

    inv_distances = Nx.new_axis(1 / (1 + distances), 2)

    grads = 4 * (pq_diff * y_diff * inv_distances
    Nx.sum(grads, axes: [1])
```
Not sure where that `* 2` came from, though.